spec/datadog/tracing/contrib/action_pack/action_dispatch/journey/router_spec.rb

```diff
@@ -0,0 +1,249 @@
+require 'datadog/tracing/contrib/support/spec_helper'
+require 'datadog/tracing/contrib/action_pack/action_dispatch/instrumentation'
+
+require 'action_pack'
+require 'action_controller/railtie'
+require 'action_view/railtie'
+require 'rails/test_unit/railtie'
+
+require 'spec/datadog/tracing/contrib/rails/support/configuration'
+require 'spec/datadog/tracing/contrib/rails/support/application'
+
+RSpec.describe 'Datadog::Tracing::Contrib::ActionPack::ActionDispatch::Journey::Router' do
+  let(:no_db) { true }
+
+  include Rack::Test::Methods
+  include_context 'Rails test application'
+
+  after do
+    Datadog.configuration.tracing[:action_pack].reset!
+    Datadog.registry[:rack].reset_configuration!
+  end
+
+  describe '#find_routes' do
+    before do
+      engine.routes.append do
+        get '/sign-in' => 'tokens#create'
+      end
+
+      auth_engine = engine
+      rack_status_app = rack_app.new
+
+      rails_test_application.instance.routes.append do
+        namespace :api, defaults: { format: :json } do
+          resources :users, only: %i[show]
+
+          mount auth_engine => '/auth'
+
+          match '/status', to: rack_status_app, via: :get
+        end
+
+        get '/items/:id', to: 'items#by_id', id: /\d+/
+        get '/items/:slug', to: 'items#by_slug', id: /(\w-)+/
+
+        get 'books(/:category)', to: 'books#index'
+        get 'books/*section/:title', to: 'books#show'
+      end
+    end
+
+    let(:rack_app) do
+      stub_const(
+        'RackStatusApp',
+        Class.new do
+          def call(_env)
+            [200, { 'Content-Type' => 'text/plain' }, ['OK']]
+          end
+        end
+      )
+    end
+
+    let(:controllers) { [users_controller, items_controller, books_controller] }
+
+    let(:users_controller) do
+      stub_const(
+        'Api::UsersController',
+        Class.new(ActionController::Base) do
+          def show
+            head :ok
+          end
+        end
+      )
```

Comment:
I'm still new, but is there a reason we don't use `class_double` here?

https://www.rubydoc.info/github/rspec/rspec-mocks/RSpec%2FMocks%2FExampleMethods:class_double

It will allow us to control stabbing of non-existent methods, but sometimes it could work bad when class has dynamic methods.

Reply:
I don't know, this is what we used in other places for creating temporary controller classes. I guess using `class_double` on `ActionController::Base` won't allow us stubbing action methods, since they don't exist on `ActionController::Base`?

Reply:
Yes you can't stub non-existent methods, but you should not mock the `Base` class, but `UsersController` instead.

But I think now I see the reason, we will not be able to return `head(:ok)` response because we will not have the context in the RSpec file (we need to be "inside" the Base class)

Sorry for the noise 🥺